### ISSUE TITLE
Make error handler in @backstage/backend-common return JSON instead of text

### ISSUE THREAD

**Issue by Rugvip** [backstage/backstage#4424]
## Feature Suggestion

The default error handler currently returns plain text, which makes it difficult to extend the response in clean ways. I suggest we move over to using the following format as the base:

```ts
type ErrorResponseBody = {
  error: {
    name: string
    message: string
    stack?: string
  }
}
```

Would could also extend it with extra fields such as `reason`, `extra`, etc.

## Possible Implementation

Switch to returning the above format, which will be a breaking change. In the frontend we tweak error response parsing to check the response content type and handle it appropriately. We could also provide some common utilities for error parsing in the frontend and backend.

**Comment by freben** [comment]
Two perhaps interesting other specs:

https://cloud.google.com/apis/design/errors#http_mapping

https://developers.google.com/search-ads/v2/standard-error-responses

The jsonapi separates out a title (not changing per occurrence) from details
https://jsonapi.org/format/#error-objects

If we intend to make use of the new WarningPanel feature of separating out a title from details, this may actually be handy.

There's also an actual RFC
https://tools.ietf.org/html/rfc7807

for which there's like https://www.npmjs.com/package/problem-json

**Comment by Rugvip** [comment]
Fixes #4424 !
